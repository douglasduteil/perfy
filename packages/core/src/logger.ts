
import { injectable } from 'inversify';

const noop = Function.prototype;

export const silentLog = {
  error: noop,
  warn: noop,
  info: noop,
  verbose: noop,
  silly: noop,
  http: noop,
  pause: noop,
  resume: noop
};

@injectable()
export abstract class Logger {
  abstract error(prefix: string | boolean, ...message: any[]): void;
  abstract warn(prefix: string | boolean, ...message: any[]): void;
  abstract info(prefix: string | boolean, ...message: any[]): void;
  abstract verbose(prefix: string | boolean, ...message: any[]): void;
  abstract silly(prefix: string | boolean, ...message: any[]): void;
  abstract http(prefix: string | boolean, ...message: any[]): void;
  abstract pause(prefix: string | boolean, ...message: any[]): void;
  abstract resume(prefix: string | boolean, ...message: any[]): void;
  abstract newItem(name: string): Logger;
}

@injectable()
// tslint:disable-next-line:max-classes-per-file
export class SilentLogger extends Logger {
  error(): void { }
  warn(): void { }
  info(): void { }
  verbose(): void { }
  silly(): void { }
  http(): void { }
  pause(): void { }
  resume(): void { }
  newItem(): Logger { return this; }
}
