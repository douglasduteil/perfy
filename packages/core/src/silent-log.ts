const noop = Function.prototype;

export const silentLog = {
  error: noop,
  warn: noop,
  info: noop,
  verbose: noop,
  silly: noop,
  http: noop,
  pause: noop,
  resume: noop
};
