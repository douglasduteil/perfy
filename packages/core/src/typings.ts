//

import Lowdb = require('lowdb');

export interface IDatabase {
  suites: {};
}

export interface IUpdateOptionType {
  cwd: string;
  database: Lowdb;
  pattern: string;
  reportsFolder: string;
}
