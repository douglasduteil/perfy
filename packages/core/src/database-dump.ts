//

import { join } from 'path';

import Lowdb = require('lowdb');
import * as fse from 'fs-extra';
import { inject, injectable, optional } from 'inversify';
import { omit } from 'lodash';

import { IPerfySuite } from '@perfyjs/types';
import { Logger } from './logger';

@injectable()
export class DatabaseDump {
  static SUITE_INDEX_FILE_NAME = Symbol('SUITE_INDEX_FILE_NAME');

  static COPY_FS = Symbol('COPY_FN');

  //

  @inject(DatabaseDump.COPY_FS)
  private fse: typeof fse;

  //
  private log: Logger;

  constructor(
    // required on instanciation
    logger: Logger,

    @inject(DatabaseDump.SUITE_INDEX_FILE_NAME)
    @optional()
    private suiteIndexFileName = 'suites.json'
  ) {
    this.log = logger.newItem('DatabaseDump');
    this.log.silly('new');
  }

  dump(database: Lowdb, destination: string) {
    const suiteIndex = join(destination, this.suiteIndexFileName);
    const suites = database.get('suites') as any;
    const suiteIndexData = suites
      .mapValues((suite: any) => omit(suite, ['cases']))
      .value();

    const suiteFolder = join(destination, 'suites');
    const writeAllSuites: Promise<void>[] = suites.map(
      (suite: IPerfySuite, suiteName: string) => this.fse.writeJson(
        join(suiteFolder, `${suiteName}.json`),
        suite
      )
    ).value();

    return Promise.all([
      this.fse.writeJson(suiteIndex, suiteIndexData),
      ...writeAllSuites
    ]);
  }
}
