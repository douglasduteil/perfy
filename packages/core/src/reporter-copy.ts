//

import * as fse from 'fs-extra';

import { inject, injectable } from 'inversify';

import { Logger } from './logger';
import { silentLog } from './silent-log';

//

export const context = {
  log: silentLog,
  copyFn: fse.copy
};

export const factory = ({log, copyFn} = context) => (from: string, to: string) => {
  log.silly('@perfy/core', 'copy', from, to);

  return Promise.all([
    copyFn(from, to)
  ]);
};

export const copy = factory();

@injectable()
export class ReporterCopy {

  static COPY_FS = Symbol('COPY_FN');

  //

  @inject(ReporterCopy.COPY_FS)
  private fse: typeof fse;

  //

  private log: Logger;
  constructor(
    // required on instanciation
    logger: Logger
  ) {
    this.log = logger.newItem('ReporterCopy');
    this.log.silly('new');
  }

  copy(from: string, to: string) {
    this.log.silly('copy', from, to);

    return Promise.all([
      this.fse.copy(from, to)
    ]);
  }
}
