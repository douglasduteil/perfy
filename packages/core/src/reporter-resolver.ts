//

import { dirname, resolve as resolvePath } from 'path';

import { inject, injectable } from 'inversify';

import { Logger } from './logger';
import { silentLog } from './silent-log';
export const context = {
  log: silentLog,
  resolveFn: require.resolve
};

export const factory = ({log, resolveFn} = context) => (name: string) => {
  log.silly('@perfy/core', 'resolve', name);

  const reporterPath = dirname(
    resolveFn(`@perfyjs/reporter-${name}`)
  );
  log.silly('@perfy/core', 'reporterPath', reporterPath);

  return resolvePath(reporterPath, 'dist');
};

export const resolve = factory();

@injectable()
export class ReporterResolver {

  static MODULE_RESOLVER = Symbol('MODULE_RESOLVER');

  //

  @inject(ReporterResolver.MODULE_RESOLVER)
  private moduleResolver: {resolve: typeof require.resolve};

  //
  private log: Logger;

  constructor(
    // required on instanciation
    logger: Logger
  ) {
    this.log = logger.newItem('ReporterResolver');
    this.log.silly('new');
  }

  resolve(name: string) {
    this.log.silly('resolve', name);

    const reporterPath = dirname(
      this.moduleResolver.resolve(`@perfyjs/reporter-${name}`)
    );
    this.log.silly('reporterPath', reporterPath);

    return resolvePath(reporterPath, 'dist');
  }
}
