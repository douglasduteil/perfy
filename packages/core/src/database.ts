//

import { inject, injectable } from 'inversify';
import Lowdb = require('lowdb');
import * as t from 'tcomb';

import { Logger } from './logger';
import { IDatabase } from './typings';

//

const defaultLowdbOptions: Lowdb.Options = {
  storage: require('lowdb/lib/storages/file-async.js'),
  writeOnChange: false
};

//

export function database(
  fileName: string,
  lowdbOptions = defaultLowdbOptions
)  {
  // log(database, fileName);

  return new Lowdb(fileName, lowdbOptions)
    .defaults<IDatabase>({ suites: {} });
}

@injectable()
export class Database {
  static FILE_NAME = Symbol('FILE_NAME');
  static LOWDB_OPTIONS = Symbol('LOWDB_OPTIONS');

  //

  @inject(Database.FILE_NAME)
  private filename: string;

  @inject(Database.LOWDB_OPTIONS)
  private lowdbOptions: Lowdb.Options;

  //

  private _instance: Lowdb;
  private log: Logger;

  constructor(
    // required on instanciation
    logger: Logger
  ) {
    this.log = logger.newItem('Database');
    this.log.silly('new');
  }

  instanciate() {
    this.log.silly('init');
    this._instance = new Lowdb(
      this.filename,
      this.lowdbOptions
    )
    .defaults<IDatabase>({ suites: {} });
  }
  get instance() {
    this.log.silly('instance');

    t.assert(Boolean(this._instance), 'No database instance found');

    return this._instance;
  }
}
