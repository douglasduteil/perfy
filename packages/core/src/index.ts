//

import * as ReporterCopy from './reporter-copy';
import * as ReporterResolver from './reporter-resolver';

export * from './database-dump';
export * from './database';
export * from './logger';
export * from './typings';
export * from './update';

export {
  ReporterResolver,
  ReporterCopy
};
