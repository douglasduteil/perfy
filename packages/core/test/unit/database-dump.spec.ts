//

import { Container } from 'inversify';
import { mock } from 'sinon';
import Lowdb = require('lowdb');

import { DatabaseDump } from 'src/database-dump';
import { Logger, SilentLogger } from 'src/logger';
import { IPerfySuites } from '@perfyjs/types';

//

const fakeFsExtra = { writeJson: Function.prototype };
const testingLowDbOption: Lowdb.Options = {
  storage: require('lowdb/lib/storages/memory.js')
};

//

test('should dump an empty suites-index.json file', async () => {
  // given
  const container = new Container();
  container.bind(DatabaseDump.COPY_FS).toConstantValue(fakeFsExtra);
  container.bind(DatabaseDump).toSelf();
  container.bind(Logger).to(SilentLogger);

  const fakeDatabase = new Lowdb('in memory db', testingLowDbOption)
    .defaults({ suites: {} });

  const databaseDump = container.get(DatabaseDump);

  // expect
  const mockFsExtra = mock(fakeFsExtra);
  const writeJsonExpect = mockFsExtra.expects('writeJson')
  writeJsonExpect.once().withArgs('dist/suites.json', {  }).resolves();

  // when
  await databaseDump.dump(fakeDatabase, 'dist');

  // verify
  expect(() => mockFsExtra.verify()).not.toThrow();
});

test.only('should dump the foo and bar suite', async () => {
  // given
  const ff = jest.fn().mockReturnValue(Promise.resolve());
  const container = new Container();
  container.bind(DatabaseDump.COPY_FS).toConstantValue({ writeJson: ff});
  container.bind(DatabaseDump).toSelf();
  container.bind(Logger).to(SilentLogger);

  const suites: IPerfySuites = {
    foo: {name: 'Foo', cases: [{id: 'foo iterations', iterations: []}]},
    bar: {name: 'Bar', cases: [{id: 'bar iterations', iterations: []}]}
  };

  const fakeDatabase = new Lowdb('in memory db', testingLowDbOption)
    .defaults({ suites });

  const databaseDump = container.get(DatabaseDump);

  // when
  await databaseDump.dump(fakeDatabase, 'dist');

  // verify
  expect(ff.mock.calls).toMatchSnapshot();
});
