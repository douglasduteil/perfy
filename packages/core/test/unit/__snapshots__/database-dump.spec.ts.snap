// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should dump the foo and bar suite 1`] = `
Array [
  Array [
    "dist/suites/foo.json",
    Object {
      "cases": Array [
        Object {
          "id": "foo iterations",
          "iterations": Array [],
        },
      ],
      "name": "Foo",
    },
  ],
  Array [
    "dist/suites/bar.json",
    Object {
      "cases": Array [
        Object {
          "id": "bar iterations",
          "iterations": Array [],
        },
      ],
      "name": "Bar",
    },
  ],
  Array [
    "dist/suites.json",
    Object {
      "bar": Object {
        "name": "Bar",
      },
      "foo": Object {
        "name": "Foo",
      },
    },
  ],
]
`;
